fuel/utils.py

```diff
@@ -14,6 +16,390 @@
+                             'the original number of examples')
+        if (slice_.start is not None and slice_.stop is not None and
+                slice_.start > slice_.stop):
+            raise ValueError('Subset instances cannot be defined by a slice '
+                             'whose start value is greater than its stop '
+                             'value')
+
+    def _request_sanity_check(self, list_or_slice, num_examples):
+        if self._is_list(list_or_slice):
+            self._list_request_sanity_check(list_or_slice, num_examples)
+        else:
+            self._slice_request_sanity_check(list_or_slice, num_examples)
+
+    def _list_request_sanity_check(self, indices, num_examples):
+        if len(indices) == 0:
+            raise ValueError('list-based requests cannot be empty (this would '
+                             'produce an empty return value)')
+        if any(index < 0 for index in indices):
+            raise ValueError('Subset does not support list-based requests '
+                             'with negative indices')
+        if max(indices) >= num_examples:
+            raise ValueError('list-based requests cannot contain indices '
+                             'greater than or equal to the number of examples '
+                             'the subset spans')
+
+    def _slice_request_sanity_check(self, slice_, num_examples):
+        numeric_args = (arg for arg in (slice_.start, slice_.stop, slice_.step)
+                        if arg is not None)
+        if any(arg < 0 for arg in numeric_args):
+            raise ValueError('Subset does not support slice-based requests '
+                             'with negative start, stop or step arguments')
+        if slice_.stop is not None and slice_.stop > num_examples:
+            raise ValueError('slice-based requests cannot have a stop value '
+                             'greater than the number of examples the subset '
+                             'spans (this would produce a return value with '
+                             'smaller length than expected')
+        if slice_.start is not None and slice_.start >= num_examples:
+            raise ValueError('slice-based requests cannot have a start value '
+                             'greater than the number of examples the subset '
+                             'spans (this would produce an empty return '
+                             'value)')
+        if (slice_.start is not None and slice_.stop is not None and
+                slice_.start >= slice_.stop):
+            raise ValueError('slice-based requests cannot have a start value '
+                             'greater than or equal to its stop value (this '
+                             'would produce an empty return value)')
+
+    def _beautify_list(self, indices):
+        # List elements should be unique and sorted
+        indices = list(sorted(set(indices)))
```

Comment:
Just for the record: this might take a while if the list is long. But we can have it as is for a while, for premature optimization is typically a bad thing.